### PR TITLE
[Design] SQLFlow Parser

### PR DESCRIPTION
**Pull request by tonyyang-svail**

### CODE REVIEW

File: doc/design/design_sqlflow_parser.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# SQLFlow Parser
+
+SQLFlow's user experience should integrate smoothly with SQL programmer's daily work routine, which involves
+
+1. Submitting multiple SQL statements at once.
+1. Working with complicated select statements.
+
+So the SQLFlow parser should
```
Comment (wangkuiyi) [inline doc/design/design_sqlflow_parser.md]: I don't get the logical connection around this "so". There are two bullets before "so" and three after it. Which one of the first two is the reason for which one in the last three?

File: doc/design/design_sqlflow_parser.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+1. Support any SQL dialect.
+1. Accept a SQL program contains multiple SQL statements.
+1. Support arbitrary select statements for extended syntax.
+
+The design of SQLFlow parser should support the above requirements.
+
+## Overview
+
+**The third-party parser** means the pre-existing parser of a particular SQL engine, like HiveQL/Calcite/TiDB parser. It supports parsing on a specific SQL dialect.
```
Comment (wangkuiyi) [inline doc/design/design_sqlflow_parser.md]: pre-existing => existing

File: doc/design/design_sqlflow_parser.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+So the SQLFlow parser should
+
+1. Support any SQL dialect.
+1. Accept a SQL program contains multiple SQL statements.
+1. Support arbitrary select statements for extended syntax.
+
+The design of SQLFlow parser should support the above requirements.
+
+## Overview
+
```
Comment (wangkuiyi) [inline doc/design/design_sqlflow_parser.md:6]: You skip over the logical progression here and enumerates some concepts. This is confusing -- why do you list these concepts.

1. We want SQLFlow to work with existing SQL engines, while extends them to support AI.
2. We want the extension is in the form of appending SQLFlow extension clauses to a SQL SELECT statement.
3. Each SQL engine might have its own SQL dialect of the SELECT statement.
4. We need the **SQLFlow parser** to call **dialect parser** to handle the SELECT statement and calls the **SQLFlow syntax extension parser** to process the appended clauses.

The logical progression reveals the concepts naturally in a logical order, which helps comprehension.

File: doc/design/design_sqlflow_parser.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+SQLFlow is written in Go, third party parsers for Calcite/HiveQL are written in Java. So we need to call the third parse via either gRPC or local command line.
+
+We prefer the local command line for the following reason.
+1. We can avoid setting up an additional gRPC server, which makes the testing/deployment simpler. And because there is no downstream call for the parser, wrapping the parser as a service might be overkill.
+1. We can guarantee the isolation between parsing requests. The gPRC server may fail due to an unexpected request, meaning one request can affect the other.
+
+## Implementation Details
+
+### Change `TRAIN/PREDICT` to `TO TRAIN/PREDICT`
```
Comment (typhoonzero) [inline doc/design/design_sqlflow_parser.md]: `TO TRAIN/PREDICT` -> `TO TRAIN/TO PREDICT` is more clear?

Comment (typhoonzero) [reply doc/design/design_sqlflow_parser.md]: What about `ANALYZE`?

**Review by typhoonzero — DISMISSED** [review]
LGTM with minor comments

**Review by wangkuiyi — COMMENTED** [review]
I am not quite sure I get the point of this design. I'd expected an algorithm showing how and when to call the SQL dialect parsers and the SQLFlow extension parser. I'd also expected a derivation from product requirement to the input/output of the parsing algorithm. However, I don't read them in this design.

File: doc/design/design_sqlflow_parser.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+Pros:
+1. Straight forward to implement.
+
+Cons:
+1. Incompleteness: single quote, double quote, square parentheses have different meanings in different SQL dialects. SQLFlow Lexer can't tokenize SQL programs of varying SQL dialects uniformly. This may lead to errors in the splitting.
+1. Needs to deal with special cases like `ALTER TABLE table_name RENAME TO train`.
+
+#### Solution 2: Splitting using the Third Party Parser
+
+SQLFlow begins parsing by using the third party parser(TPP). If TPP encounters an error at position `p`, TPP redoes parsing on the SQL statement before `p`, if success, extended SQL parser will parse on the statement after `p`. SQLFlow repeat these steps until the end of the SQL program.
```
Comment (Yancey0623) [inline doc/design/design_sqlflow_parser.md]: What to do if parsing failed before `p`?  Does it mean SQL statement is an invalid stander SQL?

File: doc/design/design_sqlflow_parser.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+SQLFlow does the following step.
+
+1. TPP parses on the first line, stop at the `;`, and succeed.
+1. TPP parses on the second line and encounters an error at `TO TRAIN`.
+    1. TPP parses the `SELECT ...` and succeeds.
+    1. The extended parser parses `TO TRAIN ...`, stops at `;` and succeeds.
+1. TPP parses on the third line and encounters an error at `TO PREDICT`.
+    1. TPP parses the `SELECT ...` and succeeds.
+    1. The extended parser parses `TO PREDICT ...`, stops at `;` and succeeds.
+1. TPP parse on the fourth line and succeed.
```
Comment (Yancey0623) [inline doc/design/design_sqlflow_parser.md]: Just a mention, for the future `VALIDATE` future: `SELECT ... TRAIN ... VALIDATE (SELECT ...)` , we also need to parse the validate SELECT statement.

File: doc/design/design_sqlflow_parser.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+1. MySQL: `TO` is used in [`GRANT`](https://dev.mysql.com/doc/refman/8.0/en/grant.html) and [`RENAME`](https://dev.mysql.com/doc/refman/8.0/en/rename-table.html).
+1. MaxCompute:
+    1. [`ALTER TABLE table_name RENAME TO new_table_name;`](https://www.alibabacloud.com/help/doc-detail/73768.html?spm=a2c5t.11065259.1996646101.searchclickresult.5afd4bd7qECSMQ)
+    1. [`alter table table_name changeowner to 'ALIYUN$xxx@aliyun.com';`](https://www.alibabacloud.com/help/doc-detail/73768.html?spm=a2c5t.11065259.1996646101.searchclickresult.5afd4bd7qECSMQ)
+
+So Given a SQL statement `SELECT ... TO TRAIN`, the third-party parser should raise error precisely at the position of `TO`.
+
+### Third Party Parser API
+
+The third-party parser provides a function `ParseAndSplit`, it takes a SQL program and returns three elements: `statements`, `position`, and `error`.
```
Comment (Yancey0623) [inline doc/design/design_sqlflow_parser.md:108]: Do the TPP API also need an API to return the AST or just columns to verify the SELECT columns and COLUMN statement 
 in extended SQL?

**Review by Yancey0623 — DISMISSED** [review]
LGTM, some tiny comment but not block merge :)

File: doc/design/design_sqlflow_parser.md (hunk 5 above)
Comment (tonyyang-svail) [reply doc/design/design_sqlflow_parser.md]: > Does it mean SQL statement is an invalid stander SQL?

Yes. Thanks for mentioning this. I will replace this paragraph with a concrete algorithm.

File: doc/design/design_sqlflow_parser.md (hunk 3 above)
Comment (tonyyang-svail) [reply doc/design/design_sqlflow_parser.md:6]: Hi @wangkuiyi, thanks for the suggestion. I have rewrote the whole section to make the logical progression more smoothly.

File: doc/design/design_sqlflow_parser.md (hunk 1 above)
Comment (tonyyang-svail) [reply doc/design/design_sqlflow_parser.md]: You are right. I will rewrite this paragraph.

File: doc/design/design_sqlflow_parser.md (hunk 7 above)
Comment (tonyyang-svail) [reply doc/design/design_sqlflow_parser.md:108]: I believe TPP only does the syntax checking. Take `select * from some_table to train ... column c1, c2` for example, the parse doesn't and shouldn't check whether `select *` has `c1`, `c2`.

File: doc/design/design_sqlflow_parser.md (hunk 4 above)
Comment (tonyyang-svail) [reply doc/design/design_sqlflow_parser.md]: > TO TRAIN/PREDICT -> TO TRAIN/TO PREDICT is more clear?

Sure.

> What about ANALYZE?

Added.

**Review by wangkuiyi — APPROVED** [review]

**Pull request merged**